server/serviceadapter.js

```diff
@@ -0,0 +1,40 @@
+class ServiceAdapter {
+  /**
+   * A string that identifies this service adapter.
+   */
+  get serviceId() {
+    if (this.constructor.SERVICE_ID == null) {
+      throw `Service adapter ${this.constructor.name} does not have a static SERVICE_ID property`;
+    }
+
+    return this.constructor.SERVICE_ID;
+  }
+
+  /**
+   * A boolean that indicates whether video metadata can be safely cached.
+   * @returns {boolean}
+   */
+  get cacheSafe() {
+    return true;
+  }
+
+  /**
+   * Returns true if this service adapter can handle a given link.
+   * @param {string} link
+   * @returns {boolean}
+   */
+  canHandleLink() {
+    return false;
+  }
+
+  /**
+   * Fetches video metadata from the API.
+   * @param {string} videoId
+   * @returns {Promise}
+   */
+  getVideoInfo(videoId, onlyProperties) {
+    return this._getVideoInfo(videoId, onlyProperties);
+  }
```

Comment:
I don't really like having `getVideoInfo` and then `_getVideoInfo` for the derived service adapters. Can we just have `getVideoInfo`?

Reply:
I don't like it either tbh. I did this to be able to utilize code completion and in-editor documentation. If `ServiceAdapter` doesn't have any reference to `getVideoInfo`, the editor won't know the method exists. If you do something like
```js
  getVideoInfo() {}
```
so that child classes can override it, the editor won't know about its parameters. The `_` version is the only way I can think of that preserves all this information.